crates/kumod/src/smtp_dispatcher.rs

```diff
@@ -55,11 +58,14 @@ impl SmtpDispatcher {
         } else {
             let mut addresses = vec![];
             for a in proto_config.mx_list.iter() {
-                addresses.append(
-                    &mut resolve_a_or_aaaa(a)
-                        .await
-                        .with_context(|| format!("resolving mx_list entry {a}"))?,
-                );
+                let addrs = &mut resolve_a_or_aaaa(a)
+                    .await
+                    .with_context(|| format!("resolving mx_list entry {a}"))?;
+                match proto_config.mx_list_ip_map.get(a) {
+                    Some(name) => addrs[0].name = name.to_string(),
+                    None => (),
+                }
+                addresses.append(addrs);
```

Comment:
Thanks for this!

How about this as an alternative implementation approach:

Add a new type:

```rust
#[derive(Deserialize, Serialize, Debug, Clone, Default)]
#[serde(untagged)]
pub enum MxListEntry {
   // A name that needs to be resolved to its A or AAAA record in DNS,
   // or an IP domain literal enclosed in square brackets like `[10.0.0.1]`
   Name(String),
   // A pre-resolved name and IP address
   Resolved(ResolvedAddress)
}
```

The `untagged` attribute will allow serde to pattern match either a string or a lua table to this type,
provided the lua table matches the shape of the `ResolvedAddress` type defined in `kumo-log-types`.

Change the `mx_list` field in `SmtpProtocol` to use this type instead of just `String`:

```rust
    #[serde(default)]
    pub mx_list: Vec<MxListEntry>,   
```

Change this section of code to resolve if needed and append to the list:

```suggestion
                match a {
                    MxListEntry::Name(a) => {
                            addresses.append(&mut resolve_a_or_aaaa(a)
                               .await
                               .with_context(|| format!("resolving mx_list entry {a}"))?));
                    }
                    MxListEntry::Resolved(addr) => {
                            addresses.append(addr);
                     }
                }
```

Now, `mx_list` can be specified like:

```lua
   mx_list = { 'smart.host.local', {name="mx.example.com", addr='10.0.0.1'} },
```

and we won't need to add the additional `mx_list_ip_map`.

Reply:
this is way better, I'll push a new commit with these changes